### PR TITLE
Better API for estimatesmartfee RPC 

### PR DESCRIPTION
**Pull request by morcos**
Through 0.14 branch, the estimatesmartfee API was tagged "WARNING: This interface is unstable and may disappear or change!" and this warning is removed for 0.15, so any wanted API updates should happen now.

The changes here are to make the additional parameter for conservative estimates a more general estimate_mode string , to omit the feerate and include an error string instead of returning -1 on error, and to do better parameter checking initially.

~It is only the last 2 commits, but it's built on #10706 and #10543~.

See https://github.com/bitcoin/bitcoin/pull/10707#issuecomment-314869251 for renaming of nblocks argument to conf_target.  Will need to be included before string freeze.

PR description edited for clarity

### CODE REVIEW

File: src/rpc/mining.cpp [hunk 1]
```diff
@@ -836,36 +836,56 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
             "for which the estimate is valid. Uses virtual transaction size as defined\n"
             "in BIP 141 (witness data is discounted).\n"
             "\nArguments:\n"
-            "1. nblocks       (numeric)\n"
-            "2. conservative  (bool, optional, default=true) Whether to return a more conservative estimate which\n"
-            "                 also satisfies a longer history. A conservative estimate potentially returns a higher\n"
-            "                 feerate and is more likely to be sufficient for the desired target, but is not as\n"
-            "                 responsive to short term drops in the prevailing fee market\n"
+            "1. nblocks         (numeric) Confirmation target in blocks (1 - 1008)\n"
+            "2. \"estimate_mode\" (string, optional, default=CONSERVATIVE) The fee estimate mode.\n"
```
Comment (morcos) [inline src/rpc/mining.cpp]: s/conservative/estimate_mode/ 7 lines up..

Comment (morcos) [inline src/rpc/mining.cpp]: done

**Comment by sipa** [comment]
Needs rebase.

**Comment by morcos** [comment]
Rebased correctly for #10589 merge

**Comment by morcos** [comment]
Clean rebase again for changes to #10707 

Added a new optional commit which changes the named arguments in `estimatesmartfee` and `estimaterawfee` from nblocks to conf_target.  `estimaterawfee` is new for 0.15 and `estimatesmartfee` was previously labeled unstable.  I think it makes sense to align the argument names here with the ones used in `fundrawtransaction`, `sendtoaddress`, and `sendmany`.  Those 3 wallet RPC calls only got this new argument for 0.15, but I think conf_target is far superior to nblocks.  Note that `bumpfee` was already released with confTarget however.  `estimatefee` is left unchanged as it is deprecated.

I don't feel strongly about this change, but if we want to align the argument names, now is the time to do it.

**Comment by ryanofsky** [comment]
> It is only the last commit, but it's built on [...]

Looks like it's the last two commits now, for anyone else looking at this.

**Review by ryanofsky — COMMENTED** [review]
utACK 7d70bb734c6ff31248a0457f47520d6e8c3fbf2f. Should update PR description to mention conf_target renames before this is merged.

**Comment by morcos** [comment]
rebased on new #10706, no changes

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -806,42 +806,61 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
...
+            "                   Whether to return a more conservative estimate which also satisfies\n"
+            "                   a longer history. A conservative estimate potentially returns a\n"
+            "                   higher feerate and is more likely to be sufficient for the desired\n"
+            "                   target, but is not asresponsive to short term drops in the\n"
```
Comment (sipa) [inline src/rpc/mining.cpp]: Typo: asresponsive

File: src/rpc/mining.cpp [hunk 3]
```diff
@@ -806,42 +806,61 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
...
 
     UniValue result(UniValue::VOBJ);
+    UniValue errors(UniValue::VARR);
     FeeCalculation feeCalc;
     CFeeRate feeRate = ::feeEstimator.estimateSmartFee(nBlocks, &feeCalc, conservative);
-    result.push_back(Pair("feerate", feeRate == CFeeRate(0) ? -1.0 : ValueFromAmount(feeRate.GetFeePerK())));
+    if (!(feeRate == CFeeRate(0))) {
+        result.push_back(Pair("feerate", ValueFromAmount(feeRate.GetFeePerK())));
+    }
+    else {
```
Comment (sipa) [inline src/rpc/mining.cpp]: Nit: else on the same line as `}`

File: src/rpc/mining.cpp [hunk 4]
```diff
@@ -806,42 +806,61 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
...
+            "                   target, but is not asresponsive to short term drops in the\n"
+            "                   prevailing fee market.  Must be one of:\n"
+            "       \"UNSET\" (defaults to CONSERVATIVE)\n"
+            "       \"ECONOMICAL\"\n"
+            "       \"CONSERVATIVE\"\n"
             "\nResult:\n"
             "{\n"
-            "  \"feerate\" : x.x,     (numeric) estimate fee-per-kilobyte (in BTC)\n"
+            "  \"feerate\" : x.x,     (numeric, optional) estimate fee-per-kilobyte (in BTC)\n"
+            "  \"errors\": [ str... ] (json array of strings, optional) Errors encountered during processing\n"
```
Comment (sipa) [inline src/rpc/mining.cpp]: Why are these not RPC errors?

Comment (morcos) [inline src/rpc/mining.cpp]: The whole command is not necessarily an error. It's just an error in returning a value for that particular horizon.  It was my understanding that this was the preferred method rather than silently omitting the horizon.  Also there is no error with the request. 

Or am I misunderstanding your question?

Comment (sipa) [inline src/rpc/mining.cpp]: I see. It makes sense that no RPC error is to be returned for cases where it is not due to something the caller could know. It still seems overkill to create a whole new errors field for just a lack of available information, though. Do we expect more kinds of errors to be added later?

Comment (morcos) [inline src/rpc/mining.cpp]: I had originally thought I'd be able to distinguish between "insufficient data" and "enough data but no passing feerate" (to the extent those are really different things) but it wasn't easy.  However I do like the idea of setting the API now to be general enough that we don't have to change the API in the future if suppose we want to give an answer but also give an errors indicating part of the calculation couldn't work or something. 

Also `estimaterawfee` now is merged to master with a similar errors field.

In short, I don't know that there will be more kinds, but I also don't see much harm in adding the errors field.  If users don't care they can just ignore it right?

File: src/rpc/mining.cpp [hunk 5]
```diff
@@ -796,43 +806,58 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
...
+        if (fee_mode == FeeEstimateMode::ECONOMICAL) conservative = false;
     }
 
     UniValue result(UniValue::VOBJ);
+    UniValue errors(UniValue::VARR);
     FeeCalculation feeCalc;
-    CFeeRate feeRate = ::feeEstimator.estimateSmartFee(nBlocks, &feeCalc, ::mempool, conservative);
-    result.push_back(Pair("feerate", feeRate == CFeeRate(0) ? -1.0 : ValueFromAmount(feeRate.GetFeePerK())));
+    CFeeRate feeRate = ::feeEstimator.estimateSmartFee(conf_target, &feeCalc, conservative);
+    if (!(feeRate == CFeeRate(0))) {
```
Comment (promag) [inline src/rpc/mining.cpp]: `operator!=`.

File: src/rpc/mining.cpp (hunk 4 above)
Comment (promag) [inline src/rpc/mining.cpp]: In terms of REST, the status code can represent both situations. In this case it could be a 412 Precondition Failed with `{ "message": "Insufficient data or no feerate found" }` as body instead of a 200 with an error body. Not sure if it can be like that with RPC.

**Comment by morcos** [comment]
Rebased now that #10706 is merged.  Typo fixed in the comments, 2 white space changes and `!(feerate == CFeeRate(0))` changed to `feerate != CFeeRate(0)`

**Review by gmaxwell — APPROVED** [review]
utACK

**Review by ryanofsky — COMMENTED** [review]
utACK a9ee31af97a65232a992c5b81ce6f1115a94d117. Only changes since previous review were rebase, switching to operator != and fixing whitespace around else.

**Comment by morcos** [comment]
Added 2 more lines of RPC help.

([smartapi.ver2](https://github.com/morcos/bitcoin/commits/smartapi.ver2)) -> 06bcdb8 ([smartapi.ver2.squash](https://github.com/morcos/bitcoin/commits/smartapi.ver2.squash))

**Comment by TheBlueMatt** [comment]
utACK 06bcdb8da64502a64df03f3c89fbc6ccb72cd349

**Review by ryanofsky — COMMENTED** [review]
utACK 06bcdb8da64502a64df03f3c89fbc6ccb72cd349, only change is help text

**Review by gmaxwell — APPROVED** [review]
reACK

File: src/rpc/mining.cpp [hunk 6]
```diff
@@ -806,42 +806,59 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
 {
     if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
         throw std::runtime_error(
-            "estimatesmartfee nblocks (conservative)\n"
+            "estimatesmartfee conf_target (\"estimate_mode\")\n"
```
Comment (promag) [inline src/rpc/mining.cpp]: Nit, missing space inside `()`,

File: src/rpc/mining.cpp [hunk 7]
```diff
@@ -806,42 +806,59 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
...
-        conservative = request.params[1].get_bool();
+        FeeEstimateMode fee_mode;
+        if (!FeeModeFromString(request.params[1].get_str(), fee_mode)) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid estimate_mode parameter");
+        }
+        if (fee_mode == FeeEstimateMode::ECONOMICAL) conservative = false;
     }
 
     UniValue result(UniValue::VOBJ);
+    UniValue errors(UniValue::VARR);
```
Comment (promag) [inline src/rpc/mining.cpp]: Nit, move to `else` block below.

File: src/rpc/mining.cpp [hunk 8]
```diff
@@ -806,42 +806,59 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
...
     FeeCalculation feeCalc;
-    CFeeRate feeRate = ::feeEstimator.estimateSmartFee(nBlocks, &feeCalc, conservative);
-    result.push_back(Pair("feerate", feeRate == CFeeRate(0) ? -1.0 : ValueFromAmount(feeRate.GetFeePerK())));
+    CFeeRate feeRate = ::feeEstimator.estimateSmartFee(conf_target, &feeCalc, conservative);
+    if (feeRate != CFeeRate(0)) {
```
Comment (promag) [inline src/rpc/mining.cpp]: Suggestion for future PR, `CFeeRate` constructor argument could have default value `= 0` so `CFeeRate()` sounds more like null fee rate.

File: src/rpc/mining.cpp (hunk 6 above)
Comment (TheBlueMatt) [inline src/rpc/mining.cpp]: Since when do we add a space inside ()s?

File: src/rpc/mining.cpp (hunk 4 above)
Comment (sipa) [inline src/rpc/mining.cpp]: We have RPC error codes, but those are usually for exceptional situations, rather than to communicate a totally inevitable not-enough-data situation.

File: src/rpc/mining.cpp [hunk 9]
```diff
@@ -806,42 +806,62 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
...
+            "An error is returned if not enough transactions and blocks\n"
             "have been observed to make an estimate for any number of blocks.\n"
             "\nExample:\n"
             + HelpExampleCli("estimatesmartfee", "6")
             );
 
-    RPCTypeCheck(request.params, {UniValue::VNUM});
-
+    RPCTypeCheck(request.params, {UniValue::VNUM, UniValue::VSTR});
+    RPCTypeCheckArgument(request.params[0], UniValue::VNUM);
```
Comment (sipa) [inline src/rpc/mining.cpp]: Isn't this redudant, and already checked on the line above?

Comment (sipa) [inline src/rpc/mining.cpp]: Nevermind.

**Comment by sipa** [comment]
utACK 06bcdb8da64502a64df03f3c89fbc6ccb72cd349

**Comment by jnewbery** [comment]
> Through 0.14 branch, the estimatesmartfee API was tagged "WARNING: This interface is unstable and may disappear or change!" and this warning is removed for 0.15

This PR does not remove the `"WARNING: This interface is unstable and may disappear or change!"` warning, and it's still in master. Should the warning be removed?

**Comment by TheBlueMatt** [comment]
@jnewbery are you sure? looks like its not there on my node

**Comment by jnewbery** [comment]
> are you sure? looks like its not there on my node

My mistake. I must have been looking at the same warning in `estimaterawfee`. This warning has definitely been removed in `estimatesmartfee` in master.
